src/lib.rs

```diff
@@ -913,71 +818,38 @@ pub unsafe extern "C" fn path_readlink(
 ) -> Errno {
     let path = slice::from_raw_parts(path_ptr, path_len);
 
-    // If the user gave us a buffer shorter than `PATH_MAX`, it may not be
-    // long enough to accept the actual path. `cabi_realloc` can't fail,
-    // so instead we handle this case specially.
-    if buf_len < PATH_MAX {
-        return path_readlink_slow(fd, path, buf, buf_len, bufused);
-    }
+    State::with(|state| {
+        // If the user gave us a buffer shorter than `PATH_MAX`, it may not be
+        // long enough to accept the actual path. `cabi_realloc` can't fail,
+        // so instead we handle this case specially.
+        let use_state_buf = buf_len < PATH_MAX;
 
-    register_buffer(buf, buf_len);
-
-    let result = wasi_filesystem::readlink_at(fd, path);
-
-    let return_value = match &result {
-        Ok(path) => {
-            assert_eq!(path.as_ptr(), buf);
-            assert!(path.len() <= buf_len);
-
-            *bufused = path.len();
-            ERRNO_SUCCESS
+        if use_state_buf {
+            state.register_buffer(state.path_buf.as_ptr() as *mut _, PATH_MAX);
```

Comment:
Can this use `as_mut_ptr()`, to avoid the need to cast away a const qualifier? That looks like it'd require `path_buf` to be a `Cell`, but that looks like it'd be more accurate about what this code is doing, because it is actually mutating `state`'s state.